### ISSUE TITLE
Perplexity Calculations for TOVA Cache

### ISSUE THREAD

**Issue by kevinli573** [schwartz-lab-NLP/TOVA#5]
For the perplexity calculations, I am trying to replicate the performance reported using TOVA with 512 cache size on 4096 token splits of PG19. 

Was the perplexity calculated with one forward pass like so:
```
with torch.no_grad():
    output = model(chunk_input, past_key_values=cache)
    logits = output.logits
    shift_logits = logits[..., :-1, :].contiguous()
    shift_labels = labels[..., 1:].contiguous()
    loss = F.cross_entropy(shift_logits.view(-1, shift_logits.size(-1)), shift_labels.view(-1), reduction='none')
    average_perplexity = torch.exp(torch.mean(loss))
```

or iteratively:
```
loss = 0.0
num_tokens = 0
for token_idx in tqdm(range(0, chunk_input.size(1) - 1)):
    input_ids = chunk_input[:, token_idx:token_idx+1]
    with torch.no_grad():                
        output = model(input_ids, past_key_values=cache)
        logits = output.logits
        token_loss = loss_fn(logits.view(-1, logits.size(-1)), chunk_input[:,token_idx+1:token_idx+2].view(-1))
        loss += token_loss
        num_tokens += 1
average_perplexity = torch.exp(loss / num_tokens)
```

Thanks!

**Comment by ml-oren** [comment]
Hi Kevin, 

Thank you for asking. We perform cache reduction to the pre-defined size **after** applying the attention mechanism of each layer. Therefore, feeding the model the whole sequence at once will give you identical results to the model without using TOVA. To evaluate TOVA, you should pass the model one token at a time, just like it seems you do in your **iterative (2nd) option**.

One important implementation note: due to some assumption in the transformers library, the positioning is done incorrectly for our cache. We've fixed it for model.generate(), but to make the direct model call work properly, you should pass the positions explicitly. Hence, the model call would look like this:
`output = model(input_ids, position_ids=torch.tensor([[token_idx]]), past_key_values=cache)`

Good luck!

**Comment by kevinli573** [comment]
Great, that seems to be working! Also, for the perplexity calculations, did you average the batch perplexity or calculate overall perplexity by keeping track of running loss and tokens evaluated over the batches?

Thanks!

**Comment by hassidm** [comment]
Hi Kevin,

About the perplexity calculations: we used the overall loss (over all the batches) to calculate the perplexity (i.e.: the second option in your comment).

Thanks,
Michael

**Comment by ZiweiHe** [comment]
> Hi Kevin,
> 
> About the perplexity calculations: we used the overall loss (over all the batches) to calculate the perplexity (i.e.: the second option in your comment).
> 
> Thanks, Michael

Hi, I'm not able to reproduce the ppl results on PG19 testset. 
The code I used to calculate the ppl of one sample in PG19 is pasted below:

```nlls = list()
  past_key_values = TOVACache(multi_state_size)
  
  for idx in tqdm(range(0, seq_len - 1)):
              input_ids = encodings.input_ids[:, idx : idx + 1].to(model.device)
              with torch.no_grad():
                  outputs = model(input_ids, past_key_values=past_key_values, use_cache=True)
                  logits = outputs.logits.view(-1, model.config.vocab_size)
                  past_key_values = outputs.past_key_values
                  label = encodings.input_ids[:, idx + 1 : idx + 2].to(logits.device).view(-1)
                  token_loss = loss_fn(logits, label)
              nlls.append(token_loss)
  ppl = torch.exp(sum(nlls) / len(nlls))```

Thanks,

Ziwei

**Comment by kevinli573** [comment]
> > Hi Kevin,
> > About the perplexity calculations: we used the overall loss (over all the batches) to calculate the perplexity (i.e.: the second option in your comment).
> > Thanks, Michael
> 
> Hi, I'm not able to reproduce the results on PG19. The code I used to get the ppl of one sample in PG19 is provided below:
> 
> ```
> enable_tova_caching(model)
> past_key_values = TOVACache(multi_state_size)
> nlls = list()
> 
> for idx in tqdm(range(0, seq_len - 1)):
>     input_ids = encodings.input_ids[:, idx : idx + 1].to(model.device)
>     with torch.no_grad():
>         outputs = model(input_ids, past_key_values=past_key_values, use_cache=True)
>         logits = outputs.logits.view(-1, model.config.vocab_size)
>         past_key_values = outputs.past_key_values
>         label = encodings.input_ids[:, idx + 1 : idx + 2].to(logits.device).view(-1)
>         token_loss = loss_fn(logits, label)
>     nlls.append(token_loss)
> ppl = torch.tensor(nlls).mean().exp().item()
> ```
> 
> Is there an error in my approach? Thanks, Ziwei

You would probably have to adjust your model forward call to this: 
outputs = model(input_ids, position_ids=torch.tensor([[idx]]), past_key_values=past_key_values, use_cache=True)

**Comment by ZiweiHe** [comment]
> > > Hi Kevin,
> > > About the perplexity calculations: we used the overall loss (over all the batches) to calculate the perplexity (i.e.: the second option in your comment).
> > > Thanks, Michael
> > 
> > 
> > Hi, I'm not able to reproduce the results on PG19. The code I used to get the ppl of one sample in PG19 is provided below:
> > ```
> > enable_tova_caching(model)
> > past_key_values = TOVACache(multi_state_size)
> > nlls = list()
> > 
> > for idx in tqdm(range(0, seq_len - 1)):
> >     input_ids = encodings.input_ids[:, idx : idx + 1].to(model.device)
> >     with torch.no_grad():
> >         outputs = model(input_ids, past_key_values=past_key_values, use_cache=True)
> >         logits = outputs.logits.view(-1, model.config.vocab_size)
> >         past_key_values = outputs.past_key_values
> >         label = encodings.input_ids[:, idx + 1 : idx + 2].to(logits.device).view(-1)
> >         token_loss = loss_fn(logits, label)
> >     nlls.append(token_loss)
> > ppl = torch.tensor(nlls).mean().exp().item()
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Is there an error in my approach? Thanks, Ziwei
> 
> You would probably have to adjust your model forward call to this: outputs = model(input_ids, position_ids=torch.tensor([[idx]]), past_key_values=past_key_values, use_cache=True)

Hi Kevin, thank you for your quick reply, unfortunately it's not gonna work when the idx exceeds 4096.
